### PR TITLE
Add bouncy castle platform for tests

### PR DESCRIPTION
**Pull request by yschimke**
Ready to land, but mainly allows quantifying how broken some of the test assumptions or the BC library is.

If we don't add to CI, it will mainly catalog how broken we are with Bouncy Castle.

### CODE REVIEW

**Comment by yschimke** [comment]
```
> Task :okhttp-tls:test

okhttp3.tls.HandshakeCertificatesTest > platformTrustedCertificates FAILED
    java.lang.AssertionError:
    Expecting:
     <["C=US,O=Entrust\,",
        "C=US,ST=Arizona,L=Scottsdale,O=Starfield",
        "C=US,O=SecureTrust",
        "C=US,O=DigiCert",
        "O=Entrust.net,OU=www.entrust.net/CPS_2048",
        "C=US,O=IdenTrust,CN=IdenTrust",
        "C=JP,O=SECOM",
        "C=BM,O=QuoVadis",
        "O=TeliaSonera,CN=TeliaSonera",
        "C=US,O=thawte\,",
        "C=US,ST=Arizona,L=Scottsdale,O=GoDaddy.com\,",
        "C=US,O=GeoTrust",
        "C=US,O=VeriSign\,",
        "C=TW,O=Chunghwa",
        "C=US,O=AffirmTrust,CN=AffirmTrust",
        "C=PL,O=Unizeto",
        "OU=GlobalSign",
        "C=US,OU=www.xrampsecurity.com,O=XRamp",
        "C=SE,O=AddTrust",
        "O=Digital",
        "C=NO,O=Buypass",
        "C=FI,O=Sonera,CN=Sonera",
        "C=DE,O=D-Trust",
        "C=CH,O=SwissSign",
        "C=US,ST=New",
        "C=GB,ST=Greater",
        "C=US,O=Internet",
        "C=IE,O=Baltimore,OU=CyberTrust,CN=Baltimore",
        "C=ZA,ST=Western",
        "C=US,O=Starfield",
        "C=EU,O=AC",
        "C=BE,O=GlobalSign",
        "C=IT,L=Milan,O=Actalis",
        "C=US,ST=UT,L=Salt",
        "C=LU,O=LuxTrust",
        "C=EU,L=Madrid",
        "C=US,O=The",
        "C=DE,O=T-Systems",
        "C=FR,O=KEYNECTIS,OU=ROOT,CN=KEYNECTIS"]>
    to contain:
     <["CN=Entrust"]>
    but could not find:
     <["CN=Entrust"]>
        at okhttp3.tls.HandshakeCertificatesTest.platformTrustedCertificates(HandshakeCertificatesTest.java:148)

okhttp3.tls.HeldCertificateTest > decodeEcdsa256 FAILED
    org.junit.ComparisonFailure: expected:<... PRIVATE KEY-----
    ME[ECAQAwEwYHKoZIzj0CAQYIKoZIzj0DAQcEJzAlAgEBBCA7ODT0xhGSNn4ESj6J
    lu/GJQZoU9lDrCPeUcQ28tzOW]w==
    -----END PRIVATE...> but was:<... PRIVATE KEY-----
    ME[0CAQAwEwYHKoZIzj0CAQYIKoZIzj0DAQcEMzAxAgEBBCA7ODT0xhGSNn4ESj6J
    lu/GJQZoU9lDrCPeUcQ28tzOW6AKBggqhkjOPQMBB]w==
    -----END PRIVATE...>
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at okhttp3.tls.HeldCertificateTest.decodeEcdsa256(HeldCertificateTest.java:262)

okhttp3.tls.HandshakeCertificatesTest > clientAndServer FAILED
    java.util.concurrent.ExecutionException: org.bouncycastle.tls.TlsFatalAlert: handshake_failure(40)
        at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
        at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at okhttp3.tls.HandshakeCertificatesTest.clientAndServer(HandshakeCertificatesTest.java:103)

        Caused by:
        org.bouncycastle.tls.TlsFatalAlert: handshake_failure(40)
            at org.bouncycastle.tls.AbstractTlsServer.getSelectedCipherSuite(AbstractTlsServer.java:387)
            at org.bouncycastle.jsse.provider.ProvTlsServer.getSelectedCipherSuite(ProvTlsServer.java:291)
            at org.bouncycastle.tls.TlsServerProtocol.sendServerHelloMessage(TlsServerProtocol.java:747)
            at org.bouncycastle.tls.TlsServerProtocol.handleHandshakeMessage(TlsServerProtocol.java:162)
            at org.bouncycastle.tls.TlsProtocol.processHandshakeQueue(TlsProtocol.java:546)
            at org.bouncycastle.tls.TlsProtocol.processRecord(TlsProtocol.java:464)
            at org.bouncycastle.tls.RecordStream.readRecord(RecordStream.java:224)
            at org.bouncycastle.tls.TlsProtocol.safeReadRecord(TlsProtocol.java:687)
            at org.bouncycastle.tls.TlsProtocol.blockForHandshake(TlsProtocol.java:325)
            at org.bouncycastle.tls.TlsServerProtocol.accept(TlsServerProtocol.java:83)
            at org.bouncycastle.jsse.provider.ProvSSLSocketWrap.startHandshake(ProvSSLSocketWrap.java:606)
            at org.bouncycastle.jsse.provider.ProvSSLSocketWrap.startHandshake(ProvSSLSocketWrap.java:571)
            at okhttp3.tls.HandshakeCertificatesTest.lambda$doServerHandshake$0(HandshakeCertificatesTest.java:169)

okhttp3.tls.HandshakeCertificatesTest > keyManager FAILED
    java.security.KeyStoreException: Failed to create trust manager
        at org.bouncycastle.jsse.provider.ProvTrustManagerFactorySpi.engineInit(ProvTrustManagerFactorySpi.java:171)
        at java.base/javax.net.ssl.TrustManagerFactory.init(TrustManagerFactory.java:278)
        at okhttp3.tls.internal.TlsUtil.newTrustManager(TlsUtil.kt:61)
        at okhttp3.tls.HandshakeCertificates$Builder.build(HandshakeCertificates.kt:145)
        at okhttp3.tls.HandshakeCertificatesTest.keyManager(HandshakeCertificatesTest.java:130)

        Caused by:
        java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
            at java.base/java.security.cert.PKIXParameters.setTrustAnchors(PKIXParameters.java:200)
            at java.base/java.security.cert.PKIXParameters.<init>(PKIXParameters.java:120)
            at java.base/java.security.cert.PKIXBuilderParameters.<init>(PKIXBuilderParameters.java:104)
            at org.bouncycastle.jsse.provider.ProvX509TrustManager.<init>(ProvX509TrustManager.java:61)
            at org.bouncycastle.jsse.provider.ProvTrustManagerFactorySpi.engineInit(ProvTrustManagerFactorySpi.java:167)
            ... 4 more
```

File: okhttp/src/test/java/okhttp3/BouncyCastleTest.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+    client.newCall(request).execute().use {
+      assertThat(it.protocol).isEqualTo(Protocol.HTTP_1_1)
+      assertThat(it.handshake!!.tlsVersion).isEqualTo(TlsVersion.TLS_1_3)
+    }
+  }
+
+  private fun enableTls() {
+    // Generate a self-signed cert for the server to serve and the client to trust.
+    // can't use TlsUtil.localhost with a non OpenJSSE trust manager
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/BouncyCastleTest.kt]: I guess this is a thing for us to fix!

Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/BouncyCastleTest.kt]: Why not? Is it just implementation details leaking out?

**Comment by yschimke** [comment]
> Should we have a doc page with security providers and their statuses?

Yep - https://github.com/square/okhttp/issues/5717

**Pull request merged**